### PR TITLE
Non templated problem

### PR DESCRIPTION
**Pull request by fdarricau**
cf: https://github.com/roboptim/roboptim-core-plugin-ipopt/pull/30

### CODE REVIEW

File: src/tnlp.cc [hunk 1]
```diff
@@ -45,7 +45,7 @@ namespace roboptim
       n = static_cast<Index> (solver_.problem ().function ().inputSize ());
       m = static_cast<Index> (constraintsOutputSize ());
 
-      function_t::vector_t x (n);
+      GenericFunction<EigenMatrixSparse>::vector_t x (n);
```
Comment (bchretien) [inline src/tnlp.cc]: Why no longer using `function_t`?

File: src/tnlp.cc [hunk 2]
```diff
@@ -61,13 +61,7 @@ namespace roboptim
...
 	   it != solver_.problem ().constraints ().end (); ++it)
 	{
-	  shared_ptr<solver_t::commonConstraintFunction_t> g;
-	  if (it->which () == LINEAR)
-	    g = get<shared_ptr<linearFunction_t> > (*it);
-	  else
-	    g = get<shared_ptr<nonLinearFunction_t> > (*it);
-
-	  nnz_jac_g += g->jacobian (x).nonZeros ();
+	  nnz_jac_g += (*it)->castInto<GenericDifferentiableFunction<traits_t> >()->jacobian (x).nonZeros ();
```
Comment (bchretien) [inline src/tnlp.cc]: Wouldn't it be better to do all the casting in the initialization, and store vectors of pointers to the relevant types?

File: src/tnlp.cc (hunk 1 above)
Comment (fdarricau) [inline src/tnlp.cc]: All the typedefs are not always available as a GenericFunction<traits_t>, so it is why I removed the function_t typedef.

I guess writing something like `typedef typename GenericDifferentiableFunction<traits_t> function_t;` would be ok

Comment (bchretien) [inline src/tnlp.cc]: You should have typedefs for each supported function types (plus Function), and all the casts should be done once and for all in the initialization and stored. In the end, you should not have to use any of the templated generic type (unless this is just used once in the implementation or something like that).

Comment (fdarricau) [inline src/tnlp.cc]: That makes sense. I will update the PR.

File: src/tnlp.cc [hunk 3]
```diff
@@ -111,28 +102,24 @@ namespace roboptim
...
-	       ++it, ++constraintId)
+	  for (unsigned long count = 0; count < constraintFunctions_.size(); ++count)
 	    {
 	      LOG4CXX_TRACE
 		(logger,
 		 "Compute jacobian of constraint id = " << constraintId
 		 << "to count for non-zeros elements");
 
-	      function_t::vector_t x (n);
+	      SparseFunction::vector_t x (n);
```
Comment (bchretien) [inline src/tnlp.cc]: In this case, `function_t == SparseFunction` so why did you need this change?

File: src/tnlp.cc [hunk 4]
```diff
@@ -206,37 +187,27 @@ namespace roboptim
...
-
-      typedef solver_t::problem_t::constraints_t::const_iterator
-	citer_t;
+      Eigen::Map<const SparseFunction::vector_t> x_ (x, n);
 
       int idx = 0;
       int constraintId = 0;
-      for (citer_t it = solver_.problem ().constraints ().begin ();
-	   it != solver_.problem ().constraints ().end (); ++it)
+      for (unsigned long i = 0; i < constraintFunctions_.size(); ++i)
```
Comment (bchretien) [inline src/tnlp.cc]: Keep using an iterator here (unless you've got a really good reason not to).

File: src/tnlp.hh [hunk 5]
```diff
@@ -163,20 +163,47 @@ namespace roboptim
...
+      const twiceDifferentiableFunction_t* twiceDifferentiableCostFunction_;
+
+      /// \brief Constraints
+      std::vector<function_t*> constraintFunctions_;
+
+      /// \brief Differentiable Constraints
+      std::vector<differentiableFunction_t*> differentiableConstraintFunctions_;
+
+      /// \brief Twice Differentiable Constraints
+      std::vector<twiceDifferentiableFunction_t*> twiceDifferentiableConstraintFunctions_;
```
Comment (bchretien) [inline src/tnlp.hh]: Shouldn't these be const pointers?

File: src/tnlp.hxx [hunk 6]
```diff
@@ -137,9 +115,29 @@ namespace roboptim
 	constraints_ (),
 	jacobian_ ()
```
Comment (bchretien) [inline src/tnlp.hxx]: I don't see any new initialization here...

File: src/tnlp.hxx [hunk 7]
```diff
@@ -490,21 +478,17 @@ namespace roboptim
...
-           it != solver_.problem ().constraints ().end (); ++it)
+      for (citer_t it = twiceDifferentiableConstraintFunctions_.begin();
+           it != twiceDifferentiableConstraintFunctions_.end (); ++it)
         {
-          shared_ptr<TwiceDifferentiableFunction> g;
-          if (it->which () == LINEAR)
-            g = get<shared_ptr<linearFunction_t> > (*it);
-          else
-            g = get<shared_ptr<nonLinearFunction_t> > (*it);
+          const TwiceDifferentiableFunction* g = &*(*it);
```
Comment (bchretien) [inline src/tnlp.hxx]: Is that really necessary?

File: src/tnlp.hxx [hunk 8]
```diff
@@ -369,18 +367,14 @@ namespace roboptim
...
-	   it != solver_.problem ().constraints ().end (); ++it)
+      for (citer_t it = constraintFunctions_.begin ();
+	   it != constraintFunctions_.end (); ++it)
 	{
-	  shared_ptr<typename solver_t::commonConstraintFunction_t> g;
-	  if (it->which () == LINEAR)
-	    g = get<shared_ptr<linearFunction_t> > (*it);
-	  else
-	    g = get<shared_ptr<nonLinearFunction_t> > (*it);
+	  const function_t* g = &*(*it);
```
Comment (bchretien) [inline src/tnlp.hxx]: Is that really necessary?

Comment (fdarricau) [inline src/tnlp.hxx]: Nope. Everywhere definitions like this remain some cleaning can still be done. I will do it
